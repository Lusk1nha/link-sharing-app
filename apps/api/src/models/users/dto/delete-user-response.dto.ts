import { ApiProperty } from '@nestjs/swagger';

export class DeleteUserResponseDto {
  @ApiProperty({
    description: 'Message indicating the result of the delete operation',
    example: 'User deleted successfully',
  })
  message = 'User deleted successfully';
}
