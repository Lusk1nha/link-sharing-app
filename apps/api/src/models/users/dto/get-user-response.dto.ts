import { ApiProperty } from '@nestjs/swagger';
import { UserEntity } from '../domain/user.entity';

export class GetUserResponseDto {
  constructor(user: UserEntity) {
    this.user = user;
  }

  @ApiProperty({
    type: UserEntity,
    description: 'User details',
  })
  user: UserEntity;
}
