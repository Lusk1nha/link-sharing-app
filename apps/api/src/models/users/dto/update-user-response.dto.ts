import { ApiProperty } from '@nestjs/swagger';
import { UserEntity } from '../domain/user.entity';

export class UpdateUserResponseDto {
  constructor(user: UserEntity) {
    this.user = user;
  }

  @ApiProperty({
    description: 'The updated user entity',
    type: UserEntity,
  })
  user: UserEntity;

  @ApiProperty({
    description: 'A message indicating the update was successful',
    example: 'User updated successfully',
  })
  message: string = 'User updated successfully';
}
