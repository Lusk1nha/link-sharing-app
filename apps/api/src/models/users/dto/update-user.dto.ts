import { ApiProperty } from '@nestjs/swagger';
import { IsEmail, IsOptional, IsString, IsUrl } from 'class-validator';

export class UpdateUserDto {
  @ApiProperty({
    description: 'The user email address',
    example: 'test@example.com',
  })
  @IsEmail()
  @IsOptional()
  email?: string;
}
