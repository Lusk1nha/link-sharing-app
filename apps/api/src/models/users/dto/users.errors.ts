import { HttpStatus } from "@nestjs/common";
import { BaseHttpException } from "src/common/exceptions/exceptions.common";

export class UserNotFoundException extends BaseHttpException {
  constructor() {
    super(
      {
        message: "User not found.",
        code: "USER_NOT_FOUND",
      },
      HttpStatus.NOT_FOUND,
      "USER_NOT_FOUND",
    );
  }
}
