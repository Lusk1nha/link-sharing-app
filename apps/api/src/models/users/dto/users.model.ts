import { ApiProperty } from "@nestjs/swagger";
import { Transform } from "class-transformer";
import { UUID } from "src/common/entities/uuid/uuid.entity";
import { UUIDFactory } from "src/common/entities/uuid/uuid.factory";

export type UserDto = {
  id: string;
  active: boolean;
  createdAt: Date;
  updatedAt: Date;
};

export class UserModel {
  @ApiProperty({
    type: UUID,
    description: "Unique identifier for the user",
  })
  readonly id: UUID;

  @ApiProperty({
    type: Boolean,
    description: "Indicates if the user is active",
  })
  readonly active: boolean;

  @ApiProperty({
    type: Date,
    description: "Timestamp when the user was created",
  })
  readonly createdAt: Date;

  @ApiProperty({
    type: Date,
    description: "Timestamp when the user was last updated",
  })
  readonly updatedAt: Date;

  constructor(data: UserDto) {
    this.id = UUIDFactory.from(data.id);
    this.active = data.active;
    this.createdAt = data.createdAt;
    this.updatedAt = data.updatedAt;
  }
}
