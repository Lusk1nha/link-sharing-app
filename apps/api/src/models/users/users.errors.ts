import { HttpStatus } from '@nestjs/common';
import { BaseHttpException } from 'src/common/exceptions/base-expections.common';

export class UserNotFoundException extends BaseHttpException {
  constructor() {
    super(
      {
        message: 'User not found.',
        code: 'USER_NOT_FOUND',
        status: HttpStatus.NOT_FOUND,
      },
      HttpStatus.NOT_FOUND,
      'USER_NOT_FOUND',
    );
  }
}

export class UserAlreadyExistsException extends BaseHttpException {
  constructor() {
    super(
      {
        message: 'User with email address already exists',
        code: 'USER_ALREADY_EXISTS',
        status: HttpStatus.CONFLICT,
      },
      HttpStatus.CONFLICT,
      'USER_ALREADY_EXISTS',
    );
  }
}
