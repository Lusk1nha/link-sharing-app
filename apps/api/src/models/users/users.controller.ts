import { Body, Controller, Patch } from '@nestjs/common';
import {
  ApiBearerAuth,
  ApiOkResponse,
  ApiOperation,
  ApiTags,
} from '@nestjs/swagger';
import { UsersService } from './users.service';
import { UpdateUserDto } from './dto/update-user.dto';
import { UUIDFactory } from 'src/common/entities/uuid/uuid.factory';

import { UpdateUserResponseDto } from './dto/update-user-response.dto';

@Controller('users')
@ApiTags('Users')
@ApiBearerAuth()
export class UsersController {
  constructor(private readonly usersService: UsersService) {}

  @Patch('update')
  @ApiOperation({ summary: 'Update user information' })
  @ApiOkResponse({ type: UpdateUserResponseDto })
  async updateUser(@Body() body: UpdateUserDto) {
    const userId = UUIDFactory.from('');

    const user = await this.usersService.updateUser(userId, body);

    return new UpdateUserResponseDto(user);
  }
}
