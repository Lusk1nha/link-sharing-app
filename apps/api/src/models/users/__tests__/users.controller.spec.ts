import { Test, TestingModule } from '@nestjs/testing';
import { UsersController } from '../users.controller';
import { UsersService } from '../users.service';
import { PrismaService } from 'src/common/database/database.service';
import { RolesModule } from 'src/models/roles/roles.module';
import { SessionsModule } from 'src/models/sessions/sessions.module';
import { CACHE_MANAGER, CacheModule } from '@nestjs/cache-manager';
import { ConfigModule } from '@nestjs/config';
import { generateSingleMockUser } from '../__mock__/users.mock';
import { JwtStoredPayload } from 'src/common/auth/__types__/auth.types';
import { Role } from 'src/common/roles/roles.common';
import { UserMapper } from '../domain/user.mapper';
import { GetUserResponseDto } from '../dto/get-user-response.dto';
import { ForbiddenResourceException } from 'src/common/auth/auth-common.errors';
import { DeleteUserResponseDto } from '../dto/delete-user-response.dto';
import { fa, faker } from '@faker-js/faker/.';
import { UpdateUserResponseDto } from '../dto/update-user-response.dto';
import { UpdateUserDto } from '../dto/update-user.dto';
import { ValidationPipe } from '@nestjs/common';

describe(UsersController.name, () => {
  let controller: UsersController;
  let usersService: UsersService;
  let prismaService: PrismaService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [
        RolesModule,
        SessionsModule,
        CacheModule.register({
          isGlobal: true,
        }),
        ConfigModule.forRoot({
          isGlobal: true,
        }),
      ],
      controllers: [UsersController],
      providers: [
        {
          provide: UsersService,
          useValue: {
            findByIdOrThrow: jest.fn(),
            updateUser: jest.fn(),
          },
        },
      ],
    }).compile();

    controller = module.get<UsersController>(UsersController);
    usersService = module.get<UsersService>(UsersService);
    prismaService = module.get<PrismaService>(PrismaService);
  });

  it('should be defined', () => {
    expect(controller).toBeDefined();
    expect(usersService).toBeDefined();
    expect(prismaService).toBeDefined();
  });

  describe('getCurrentUser route', () => {
    it(`should be defined ${UsersController.prototype.getCurrentUser.name}`, () => {
      expect(controller.getCurrentUser).toBeDefined();
    });

    it(`should call ${UsersService.prototype.findByIdOrThrow.name} when route is called`, async () => {
      const mockUser = generateSingleMockUser();
      const jwtUserPayload: JwtStoredPayload = {
        sub: mockUser.id,
        email: mockUser.email,
        roles: [Role.User],
        aud: 'api',
        iat: Date.now(),
        exp: Date.now() + 3600,
        iss: 'test-issuer',
      };

      const mockUserEntity = UserMapper.toDomain(mockUser);

      jest
        .spyOn(usersService, 'findByIdOrThrow')
        .mockResolvedValue(mockUserEntity);

      const response = await controller.getCurrentUser(jwtUserPayload);

      expect(usersService.findByIdOrThrow).toHaveBeenCalledWith(
        mockUserEntity.id,
      );
      expect(usersService.findByIdOrThrow).toHaveBeenCalledTimes(1);
      expect(response.user).toEqual(mockUserEntity);
      expect(response).toBeInstanceOf(GetUserResponseDto);
    });
  });

  describe('getUserById route', () => {
    it(`should be defined ${UsersController.prototype.getUserById.name}`, () => {
      expect(controller.getUserById).toBeDefined();
    });

    it(`should call different user if the request user is Admin`, async () => {
      const requestUser = generateSingleMockUser();
      const userToGet = generateSingleMockUser();

      const jwtUserPayload: JwtStoredPayload = {
        sub: requestUser.id,
        email: requestUser.email,
        roles: [Role.Admin],
        aud: 'api',
        iat: Date.now(),
        exp: Date.now() + 3600,
        iss: 'test-issuer',
      };

      const mockUserEntity = UserMapper.toDomain(userToGet);

      jest
        .spyOn(usersService, 'findByIdOrThrow')
        .mockResolvedValue(mockUserEntity);

      const response = await controller.getUserById(
        jwtUserPayload,
        mockUserEntity.id,
      );

      expect(usersService.findByIdOrThrow).toHaveBeenCalledWith(
        mockUserEntity.id,
      );
      expect(usersService.findByIdOrThrow).toHaveBeenCalledTimes(1);
      expect(response.user).toEqual(mockUserEntity);
      expect(response).toBeInstanceOf(GetUserResponseDto);
    });

    it(`should throw ForbiddenResourceException if the request user is not Admin`, async () => {
      const requestUser = generateSingleMockUser();
      const userToGet = generateSingleMockUser();

      const jwtUserPayload: JwtStoredPayload = {
        sub: requestUser.id,
        email: requestUser.email,
        roles: [Role.User],
        aud: 'api',
        iat: Date.now(),
        exp: Date.now() + 3600,
        iss: 'test-issuer',
      };

      const mockUserEntity = UserMapper.toDomain(userToGet);

      jest
        .spyOn(usersService, 'findByIdOrThrow')
        .mockResolvedValue(mockUserEntity);

      await expect(
        controller.getUserById(jwtUserPayload, mockUserEntity.id),
      ).rejects.toThrow(new ForbiddenResourceException());
      expect(usersService.findByIdOrThrow).toHaveBeenCalledWith(
        mockUserEntity.id,
      );
    });

    it(`should return user data when user call himself`, async () => {
      const requestUser = generateSingleMockUser();
      const jwtUserPayload: JwtStoredPayload = {
        sub: requestUser.id,
        email: requestUser.email,
        roles: [],
        aud: 'api',
        iat: Date.now(),
        exp: Date.now() + 3600,
        iss: 'test-issuer',
      };

      const mockRequestUserEntity = UserMapper.toDomain(requestUser);

      jest
        .spyOn(usersService, 'findByIdOrThrow')
        .mockResolvedValue(mockRequestUserEntity);

      const response = await controller.getUserById(
        jwtUserPayload,
        mockRequestUserEntity.id,
      );

      expect(usersService.findByIdOrThrow).toHaveBeenCalledWith(
        mockRequestUserEntity.id,
      );
      expect(usersService.findByIdOrThrow).toHaveBeenCalledTimes(1);
      expect(response.user).toEqual(mockRequestUserEntity);
      expect(response).toBeInstanceOf(GetUserResponseDto);
    });
  });

  describe('updateCurrentUser route', () => {
    it(`should be defined ${UsersController.prototype.updateCurrentUser.name}`, () => {
      expect(controller.updateCurrentUser).toBeDefined();
    });

    it(`should validate request dto`, async () => {
      const mockUser = generateSingleMockUser();

      const jwtUserPayload: JwtStoredPayload = {
        sub: mockUser.id,
        email: mockUser.email,
        roles: [Role.User],
        aud: 'api',
        iat: Date.now(),
        exp: Date.now() + 3600,
        iss: 'test-issuer',
      };

      let target: ValidationPipe = new ValidationPipe({
        transform: true,
        whitelist: true,
      });

      const body: UpdateUserDto = {
        email: 'invalid-email', // Invalid email format
      };

      await expect(
        target.transform(body, {
          type: 'body',
          metatype: UpdateUserDto,
        }),
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it(`should call updateCurrentUser route with valid data`, async () => {
      const mockUser = generateSingleMockUser();

      const mockWithNewEmail = generateSingleMockUser({
        id: mockUser.id,
        email: faker.internet.email(),
      });

      const jwtUserPayload: JwtStoredPayload = {
        sub: mockUser.id,
        email: mockUser.email,
        roles: [Role.User],
        aud: 'api',
        iat: Date.now(),
        exp: Date.now() + 3600,
        iss: 'test-issuer',
      };

      const mockUserUpdatedEntity = UserMapper.toDomain(mockWithNewEmail);

      jest
        .spyOn(usersService, 'updateUser')
        .mockResolvedValue(mockUserUpdatedEntity);

      const response = await controller.updateCurrentUser(jwtUserPayload, {
        email: mockWithNewEmail.email,
      });

      expect(usersService.updateUser).toHaveBeenCalledWith(
        mockUserUpdatedEntity.id,
        {
          email: mockWithNewEmail.email,
        },
      );
      expect(usersService.updateUser).toHaveBeenCalledTimes(1);
      expect(response.user).toEqual(mockUserUpdatedEntity);
      expect(response).toBeInstanceOf(UpdateUserResponseDto);
    });
  });

  describe('deleteCurrentUser route', () => {
    it(`should be defined ${UsersController.prototype.deleteCurrentUser.name}`, () => {
      expect(controller.deleteCurrentUser).toBeDefined();
    });

    it(`should call ${UsersService.prototype.deleteUser.name} when route is called`, async () => {
      const mockUser = generateSingleMockUser();
      const jwtUserPayload: JwtStoredPayload = {
        sub: mockUser.id,
        email: mockUser.email,
        roles: [Role.User],
        aud: 'api',
        iat: Date.now(),
        exp: Date.now() + 3600,
        iss: 'test-issuer',
      };

      const mockUserEntity = UserMapper.toDomain(mockUser);

      jest.spyOn(usersService, 'deleteUser').mockResolvedValue();

      const response = await controller.deleteCurrentUser(jwtUserPayload);

      expect(usersService.deleteUser).toHaveBeenCalledWith(mockUserEntity.id);
      expect(usersService.deleteUser).toHaveBeenCalledTimes(1);
      expect(response).toBeDefined();
      expect(response).toBeInstanceOf(DeleteUserResponseDto);
    });
  });
});
