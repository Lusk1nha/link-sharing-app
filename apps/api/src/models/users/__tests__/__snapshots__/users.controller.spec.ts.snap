// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersController updateCurrentUser route should validate request dto 1`] = `"Bad Request Exception"`;
